iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java

```diff
@@ -929,9 +934,64 @@ public DynamicPartitionCtx createDPContext(
       addCustomSortExpr(table, hmsTable, writeOperation, dpCtx, getSortTransformSpec(table));
     }
 
+    // Even if table has no explicit sort order, honor z-order if configured
+    Map<String, String> props = table.properties();
+    if ("ZORDER".equalsIgnoreCase(props.getOrDefault(SORT_ORDER, ""))) {
+      createZOrderCustomSort(props, dpCtx, table, hmsTable, writeOperation);
+    }
+
     return dpCtx;
   }
 
+  /**
+   * Adds a custom sort expression to the DynamicPartitionCtx that performs local Z-ordering on write.
+   *
+   * Behavior:
+   * - Reads Z-order properties from 'sort.order' and 'sort.columns' (comma-separated).
+   * - Resolves the referenced columns to their positions in the physical row (taking into account
+   *   ACID virtual columns offset for overwrite/update operations).
+   * - Configures a single ASC sort key with NULLS FIRST and injects a custom key expression for
+   *   Z-order
+   */
+  private void createZOrderCustomSort(Map<String, String> props, DynamicPartitionCtx dpCtx, Table table,
+          org.apache.hadoop.hive.ql.metadata.Table hmsTable, Operation writeOperation) {
+    String colsProp = props.get(SORT_COLUMNS);
+    if (StringUtils.isNotBlank(colsProp)) {
+      List<String> zCols = Arrays.stream(colsProp.split(",")).map(String::trim)
```

Comment:
List<String> zCols = Arrays.stream(colsProp.split(","))
    .map(String::trim)
    .filter(Predicate.not(String::isEmpty))
    .toList();